### PR TITLE
Refactor library and update to the lastest radon version

### PR DESCRIPTION
**Pull request by Tommytrg**

### CODE REVIEW

File: dist/radon.d.ts [hunk 1]
```diff
@@ -1,32 +1,37 @@
-import { Markup, Mir, MirScript, MirOperator, OperatorInfo, OperatorCode, MirArgument, MarkupOption, MarkupSelectedOption, MarkupInput, FilterArgument, OutputType, Reducer, CacheRef, MarkupSource, MarkupScript, MarkupOperator, MarkupArgument, CachedMarkupSelectedOption, CachedArgument, CachedMarkup, CachedMarkupSelect, CachedMarkupScript, CachedMarkupOperator } from './types';
+import { Markup, Mir, MarkupInput, Reducer, Filter, CachedMarkupSelectedOption, CachedArgument, CachedMarkup, CachedMarkupSelect, OperatorName } from './types';
 export declare class Radon {
     private cache;
+    private lasType;
```
Comment (tmpolaczyk) [inline dist/radon.d.ts]:
```suggestion
    private lastType;
```

File: src/radon.ts [hunk 2]
```diff
@@ -1,423 +1,527 @@
...
-    return markupOperator
+    if (index && this.operators[index + 1]) {
+      if (!areValidConsecutiveOperators(this.operators, index)) {
+        removeInvalidOperators(index)
+      }
+    } else if (!index) {
+      index = index
+        ? index
+        : this.operators.reduce((acc, _operator: Operator, i: number) => {
+            return acc > 0 ? acc : areValidConsecutiveOperators(this.operators, i) ? -1 : i
```
Comment (tmpolaczyk) [inline src/radon.ts]: Nice double ternary, but it's more clear to use a few ifs

File: src/radon.ts [hunk 3]
```diff
@@ -1,423 +1,527 @@
...
+  public default: Boolean
+  public scriptId: number
+  public inputType: OutputType
+  public id: number
+  public eventEmitter: EventEmitter
+
+  constructor(
+    cache: Cache,
+    scriptId: number,
+    inputType: OutputType | null,
```
Comment (tmpolaczyk) [inline src/radon.ts:266]: Does JavaScript have default arguments?

    inputType: OutputType or DEFAULT_INPUT_TYPE

File: src/radon.ts [hunk 4]
```diff
@@ -1,423 +1,527 @@
...
+    if (args.label || args.code) {
+      const operatorCode: OperatorCode = args.code
+        ? args.code
+        : getOperatorCodeFromOperatorName(args.label as OperatorName)
+      const operatorInfo = operatorInfos[operatorCode]
+      const defaultOperatorArguments = operatorInfo.arguments.map((argument: ArgumentInfo) => {
+        return getDefaultMirArgumentByType(argument.type)
+      })
+
+      this.default = false
```
Comment (tmpolaczyk) [inline src/radon.ts]: this.default is set to false here but it can never be set to true again

File: src/radon.ts [hunk 5]
```diff
@@ -1,423 +1,527 @@
...
 
-    return markupOperator
+    if (index && this.operators[index + 1]) {
+      if (!areValidConsecutiveOperators(this.operators, index)) {
+        removeInvalidOperators(index)
+      }
+    } else if (!index) {
+      index = index
+        ? index
+        : this.operators.reduce((acc, _operator: Operator, i: number) => {
```
Comment (tmpolaczyk) [inline src/radon.ts]: This should be a for loop with a break

File: src/radon.ts [hunk 6]
```diff
@@ -1,423 +1,527 @@
...
-      : {
-          code: operator as OperatorCode,
-          args: null,
-        }
+function areValidConsecutiveOperators(operators: Array<Operator>, idx: number) {
+  if (operators[idx + 1]) {
+    const outputType = operators[idx].operatorInfo.outputType
+    const label = operators[idx + 1].operatorInfo.name
+    const options = markupOptions[outputType]
+    return !options.find(operatorName => operatorName === label)
```
Comment (tmpolaczyk) [inline src/radon.ts:429]: Replace with a constant time lookup

File: src/radon.ts [hunk 7]
```diff
@@ -1,423 +1,527 @@
...
+      return OperatorCode.IntegerAbsolute
+    case Type.Map:
+      return OperatorCode.MapEntries
+    case Type.String:
+      return OperatorCode.StringAsBoolean
+  }
+}
+
+function isArrayType(type: OutputType) {
+  return (
```
Comment (tmpolaczyk) [inline src/radon.ts]: In python it's idiomatic to do

    [OutputType.Array, OutputType.ArrayArray, ...].contains(type)

Comment (tmpolaczyk) [reply src/radon.ts]: But in javascript it's probably more idiomatic to just

    type.toString().startsWith("Array")

File: src/radon.ts [hunk 8]
```diff
@@ -1,423 +1,527 @@
...
-      const cachedOperator: CachedMarkupOperator = (this.unwrapResultFromCache(
-        operatorRef
-      ) as unknown) as CachedMarkupOperator
-      const operator: MarkupOperator = this.unwrapOperator(cachedOperator, operatorRef.id)
-
-      return operator
-    })
 
-    return markupScript
+  public update(args: { label?: OperatorName; code?: OperatorCode }) {
```
Comment (tmpolaczyk) [inline src/radon.ts]: What's the problem with:

```suggestion
  public update(args: OperatorName | OperatorCode ) {
```

File: src/radon.ts (hunk 3 above)
Comment (Tommytrg) [reply src/radon.ts:266]: It has but I need to know if it is default or not

File: src/radon.ts (hunk 8 above)
Comment (Tommytrg) [reply src/radon.ts]: This is an straightforward way to avoid a match between OperatorName and OperatorCode

**Pull request merged**